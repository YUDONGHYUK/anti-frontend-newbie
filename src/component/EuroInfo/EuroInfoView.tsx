import useEuroInfoViewModel from './EuroInfoViewModel';
import Loading from '../Loading/Loading';
import styles from './EuroInfoView.module.css';

export default function EuroInfoView() {
  const { euro, eurInfo, error, isLoading, isFetching, krw, onChange } =
    useEuroInfoViewModel();

  if (isLoading || isFetching) {
    return <Loading />;
  }

  if (error) {
    return <div>{error.message}</div>;
  }

  const isHigherBasePrice =
    eurInfo && eurInfo.basePrice - eurInfo.openingPrice > 0;

  return (
    <>
      {eurInfo && (
        <div className={styles.info}>
          <h1>환율기준 (1 유로)</h1>
          <h2>
            {eurInfo.basePrice}
            <span
              className={`${styles.gap} ${
                isHigherBasePrice ? styles.higher : styles.lower
              }`}
            >
              <span>
                {isHigherBasePrice ? '▲' : '▼'}
                {eurInfo.changePrice}원
              </span>
              ({((eurInfo.changePrice / eurInfo.basePrice) * 100).toFixed(2)}%)
            </span>
          </h2>
          <ul className={styles.exchange}>
            <li>살때 : {eurInfo.cashBuyingPrice}</li>
            <li>팔때 : {eurInfo.cashSellingPrice}</li>
            <li>보낼때 : {eurInfo.ttSellingPrice}</li>
            <li>받을때 : {eurInfo.ttBuyingPrice}</li>
          </ul>
          <hr />
          <input
            type='text'
            placeholder='0'
            value={euro}
            onChange={onChange}
          />{' '}
          유로 ▶︎ <input disabled value={krw} /> 원
        </div>
      )}
    </>
  );
}
