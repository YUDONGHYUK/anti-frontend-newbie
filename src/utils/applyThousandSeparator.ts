export function applyThousandSeparator(str: string) {
  const thousandReg = /\B(?=(\d{3})+(?!\d))/g;

  return str.replace(thousandReg, ',');
}
