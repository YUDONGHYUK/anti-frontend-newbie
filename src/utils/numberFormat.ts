export function applyThousandSeparator(str: string) {
  const thousandReg = /\B(?=(\d{3})+(?!\d))/g;

  return str.replace(thousandReg, ',');
}

export function exchangeEurToKrw(euro: string, cashBuyingPrice: number = 0) {
  const numEuro = Number(euro.replaceAll(',', ''));
  const kwr = (numEuro * cashBuyingPrice).toFixed(0);

  return applyThousandSeparator(kwr);
}

export function splitDecimal(numStr: string) {
  const parts = numStr.split('.');
  const integer = parts[0];
  const decimal = parts[1] || '';

  return { integer, decimal };
}

export function checkDecimalScale(euroStr: string) {
  const regex = /^\d*.?\d{0,2}$/;

  return regex.test(euroStr);
}

export function convertOnlyNumOrDot(str: string) {
  return str.replace(/(^0+)/, '').replace(/[^0-9.]/g, '');
}
