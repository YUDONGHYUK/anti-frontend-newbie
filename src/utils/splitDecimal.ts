export function splitDecimal(numStr: string) {
  const parts = numStr.split('.');
  const integer = parts[0];
  const decimal = parts[1] || '';

  return { integer, decimal };
}
