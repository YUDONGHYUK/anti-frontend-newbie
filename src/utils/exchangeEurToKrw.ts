import { applyThousandSeparator } from './applyThousandSeparator';

export function exchangeEurToKrw(euro: string, cashBuyingPrice: number = 0) {
  const numEuro = Number(euro.replaceAll(',', ''));
  const kwr = (numEuro * cashBuyingPrice).toFixed(0);

  return applyThousandSeparator(kwr);
}
