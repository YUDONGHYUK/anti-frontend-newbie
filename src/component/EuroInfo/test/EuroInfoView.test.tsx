import renderer from 'react-test-renderer';
import EuroInfoView from '../EuroInfoView';
import useEuroInfoViewModel from '../EuroInfoViewModel';
import { withAllProviders } from '../../../utils/tests/utils';

const mockedUseEuroInfoViewModel = useEuroInfoViewModel as jest.Mock;
jest.mock('../EuroInfoViewModel');

describe('EuroInfoView', () => {
  let fakeApi: { euro: jest.Mock };

  beforeEach(() => {
    fakeApi = {
      euro: jest.fn(),
    };
  });

  afterEach(() => fakeApi.euro.mockReset());

  it('renders correctly while is loading', () => {
    mockedUseEuroInfoViewModel.mockReturnValue({
      euro: '',
      krw: '0',
      eurInfo: null,
      isLoading: true,
      isFetching: true,
      onChange: jest.fn(),
    });
    const tree = renderer
      .create(withAllProviders(<EuroInfoView />, fakeApi))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly while is fetching', () => {
    mockedUseEuroInfoViewModel.mockReturnValue({
      euro: '',
      krw: '0',
      eurInfo: null,
      isLoading: false,
      isFetching: true,
      onChange: jest.fn(),
    });
    const tree = renderer
      .create(withAllProviders(<EuroInfoView />, fakeApi))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly', async () => {
    mockedUseEuroInfoViewModel.mockReturnValue({
      euro: '1',
      krw: '1300',
      eurInfo: {
        basePrice: 1339.98,
        cashBuyingPrice: 1366.64,
        cashSellingPrice: 1313.32,
        changePrice: 3.86,
        openingPrice: 1336.18,
        ttBuyingPrice: 1326.59,
        ttSellingPrice: 1353.37,
      },
      isLoading: false,
      isFetching: false,
      onChange: jest.fn(),
    });

    const tree = renderer
      .create(withAllProviders(<EuroInfoView />, fakeApi))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
