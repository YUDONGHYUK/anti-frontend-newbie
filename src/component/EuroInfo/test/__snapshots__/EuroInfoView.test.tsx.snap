// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuroInfoView render correctly when errors occur 1`] = `
<div>
  Somethig is wrong
</div>
`;

exports[`EuroInfoView renders correctly 1`] = `
<div
  className="info"
>
  <h1>
    환율기준 (1 유로)
  </h1>
  <h2>
    1339.98
    <span
      className="gap higher"
    >
      <span>
        ▲
        3.86
        원
      </span>
      (
      0.29
      %)
    </span>
  </h2>
  <ul
    className="exchange"
  >
    <li>
      살때 : 
      1366.64
    </li>
    <li>
      팔때 : 
      1313.32
    </li>
    <li>
      보낼때 : 
      1353.37
    </li>
    <li>
      받을때 : 
      1326.59
    </li>
  </ul>
  <hr />
  <input
    onChange={[MockFunction]}
    placeholder="0"
    type="text"
    value="1"
  />
   
  유로 ▶︎ 
  <input
    disabled={true}
    value="1300"
  />
   원
</div>
`;

exports[`EuroInfoView renders correctly when baseprice < openingPrice 1`] = `
<div
  className="info"
>
  <h1>
    환율기준 (1 유로)
  </h1>
  <h2>
    1336.18
    <span
      className="gap lower"
    >
      <span>
        ▼
        3.86
        원
      </span>
      (
      0.29
      %)
    </span>
  </h2>
  <ul
    className="exchange"
  >
    <li>
      살때 : 
      1366.64
    </li>
    <li>
      팔때 : 
      1313.32
    </li>
    <li>
      보낼때 : 
      1353.37
    </li>
    <li>
      받을때 : 
      1326.59
    </li>
  </ul>
  <hr />
  <input
    onChange={[MockFunction]}
    placeholder="0"
    type="text"
    value="1"
  />
   
  유로 ▶︎ 
  <input
    disabled={true}
    value="1300"
  />
   원
</div>
`;

exports[`EuroInfoView renders correctly when baseprice > openingPrice 1`] = `
<div
  className="info"
>
  <h1>
    환율기준 (1 유로)
  </h1>
  <h2>
    1339.98
    <span
      className="gap higher"
    >
      <span>
        ▲
        3.86
        원
      </span>
      (
      0.29
      %)
    </span>
  </h2>
  <ul
    className="exchange"
  >
    <li>
      살때 : 
      1366.64
    </li>
    <li>
      팔때 : 
      1313.32
    </li>
    <li>
      보낼때 : 
      1353.37
    </li>
    <li>
      받을때 : 
      1326.59
    </li>
  </ul>
  <hr />
  <input
    onChange={[MockFunction]}
    placeholder="0"
    type="text"
    value="1"
  />
   
  유로 ▶︎ 
  <input
    disabled={true}
    value="1300"
  />
   원
</div>
`;

exports[`EuroInfoView renders correctly while is fetching 1`] = `
<div>
  Loading...
</div>
`;

exports[`EuroInfoView renders correctly while is loading 1`] = `
<div>
  Loading...
</div>
`;
